spec.emu

```diff
@@ -168,3 +168,334 @@ contributors: Surma, Dan Ehrenberg
+                  1. Append _ee_ to _localExportEntries_.
+                1. Else,
+                  1. Let _ie_ be the element of _importEntries_ whose [[LocalName]] is the same as _ee_.[[LocalName]].
+                  1. If _ie_.[[ImportName]] is ~namespace-object~, then
+                    1. NOTE: This is a re-export of an imported module namespace object.
+                    1. Append _ee_ to _localExportEntries_.
+                  1. Else,
+                    1. NOTE: This is a re-export of a single name.
+                    1. Append the ExportEntry Record { [[ModuleRequest]]: _ie_.[[ModuleRequest]], [[ImportName]]: _ie_.[[ImportName]], [[LocalName]]: *null*, [[ExportName]]: _ee_.[[ExportName]] } to _indirectExportEntries_.
+              1. Else if _ee_.[[ImportName]] is ~all-but-default~, then
+                1. Assert: _ee_.[[ExportName]] is *null*.
+                1. Append _ee_ to _starExportEntries_.
+              1. Else,
+                1. Append _ee_ to _indirectExportEntries_.
+            1. Let _async_ be _body_ Contains `await`.
+            1. Return Source Text Module Record { [[Realm]]: _realm_, [[Environment]]: ~empty~, [[Namespace]]: ~empty~, [[CycleRoot]]: ~empty~, [[HasTLA]]: _async_, [[AsyncEvaluation]]: *false*, [[TopLevelCapability]]: ~empty~, [[AsyncParentModules]]: &laquo; &raquo;, [[PendingAsyncDependencies]]: ~empty~, [[Status]]: ~unlinked~, [[EvaluationError]]: ~empty~, [[HostDefined]]: _hostDefined_, [[ECMAScriptCode]]: _body_, [[Context]]: ~empty~, [[ImportMeta]]: ~empty~, [[RequestedModules]]: _requestedModules_, [[ImportEntries]]: _importEntries_, [[LocalExportEntries]]: _localExportEntries_, [[IndirectExportEntries]]: _indirectExportEntries_, [[StarExportEntries]]: _starExportEntries_, [[DFSIndex]]: ~empty~, [[DFSAncestorIndex]]: ~empty~ }.
+          </emu-alg>
+        </emu-clause>
+      </emu-clause>
+
+      <emu-clause id="sec-resolveimportedmodule" type="abstract operation">
+        <h1>
+          <ins>
+            ResolveImportedModule (
+              _referencingScriptOrModule_: a Script Record, a Module Record, or *null*,
+              _specifier_: a String or an Object with a [[ModuleBlockBody]] internal slot,
+            ): either a normal completion containing a Module Record or a throw completion
+          </ins>
+        </h1>
+        <dl class="header"></dl>
+
+        <emu-alg>
+          1. Assert: If Type(_specifier_) is Object, then _referencingScriptOrModule_ is *null*.
+          1. Let _currentRealm_ be the current Realm.
+          1. If there exists a Record _record_ in _currentRealm_.[[ModuleMap]] such that _record_.[[ReferencingScriptOrModule]] is _referencingScriptOrModule_ and _record_.[[Specifier]] is _specifier_, then
+            1. Return NormalCompletion(_record_.[[ModuleRecord]]).
+          1. Assert: Type(_specifier_) is String.
+          1. Let _completion_ be Completion(HostResolveImportedModule(_referencingScriptOrModule_, _specifier_)).
```

Comment:
I wonder if we could also tighten the async requirements here. There's an implicit assumption that `HostResolveImportedModule` can do promise work. Now that the cache case is reified, we could even refactor the calls that are known to be cached (ie in evaluation) to clearly never result in userland hooks and never result in promise work.

Reply:
Can it? It's specified to be synchronous (it either throws, or returns a module record): https://tc39.es/ecma262/#sec-hostresolveimportedmodule

Reply:
`HostResolveImportedModule` is async in instantiate (by definition - you cannot resolve a module until you've loaded the parent, and you cannot return a module record until you've loaded the resolved module which is on the web an async network operation) and sync in execute, and this is an implicit assumption in the spec that has never been formalized. `HostImportModuleDynamcially` is only for dynamic `import()`, although I was never quite sure why it needed its own host hook.

Reply:
Unless we are implicitly assuming in ECMA-262 today that `HostResolveImportedModule` in instantiate is sync because the fetch maps are already populated. But that basically assumes that `HostResolveImportedModule` is not actually a resolver, but a reified resolver to begin with just representing another resolver that actually already did the work.

Reply:
Yes, that's exactly what's happening. And I don't think that it's implicit: `HostResolveImportedModule` explicitly returns a `Completion<Module Record>` and not `Completion<Promise<Module Record>>`.

As an example, the HTML spec guarantees that the whole graph has been loaded before any call to `HostResolveImportedModule`: step 6 of https://html.spec.whatwg.org/#hostresolveimportedmodule(referencingscriptormodule,-modulerequest)

> But that basically assumes that HostResolveImportedModule is not actually a resolver

It's as much a resolver as `import.meta.resolve` is on the web: both are synchronous pure transforms that depend only on the referrer, the specifier and the module map.

Reply:
Right, all loaders have an initial phase to gather all the modules before initiating the algorithm. This is interesting though - effectively all module loaders are two module loaders!